packages/amplify-migration-e2e/src/templategen.ts

```diff
@@ -0,0 +1,143 @@
+  await executeCommand(commands[0], cwd);
+  await retry(
+    () => assertStepCompletion(commands[1]),
+    (status) => status.includes(STATUS_COMPLETE) && !status.includes(STATUS_IN_PROGRESS),
+    RETRY_CONFIG,
+    (status) => status.includes(STATUS_FAILED),
+  );
+}
+
+async function executeStep2(cwd: string, commands: string[]) {
+  await executeCommand(commands[0], cwd);
+  await retry(
+    () => assertStepCompletion(commands[1]),
+    (status) => status.includes(STATUS_COMPLETE) && !status.includes(STATUS_IN_PROGRESS),
+    RETRY_CONFIG,
+    (status) => status.includes(STATUS_FAILED),
+  );
+}
+
+async function executeStep3(cwd: string, commands: string[], bucketName: string) {
+  toggleEnvVariable('BUCKET_NAME', 'SET', bucketName);
+  await executeCommand(commands[1], cwd);
+  await executeCommand(commands[2], cwd);
+  const stackRefactorId = await executeCreateStackRefactorCallCommand(commands[3], cwd);
+  toggleEnvVariable('STACK_REFACTOR_ID', 'SET', stackRefactorId);
+  await retry(
+    () => assertRefactorStepCompletion(commands[5]),
+    (status) => status.includes(STATUS_COMPLETE) && !status.includes(STATUS_IN_PROGRESS),
+    RETRY_CONFIG,
+    (status) => status.includes(STATUS_FAILED),
+  );
+  await executeCommand(commands[6], cwd);
+  await retry(
+    () => assertRefactorStepCompletion(commands[7]),
+    (status) => status.includes(STATUS_COMPLETE) && !status.includes(STATUS_IN_PROGRESS),
+    RETRY_CONFIG,
+    (status) => status.includes(STATUS_FAILED),
+  );
+}
+
+async function assertStepCompletion(command: string) {
+  const processResult = JSON.parse(await executeCommand(command));
+  return processResult.Stacks[0].StackStatus;
+}
+
+async function assertRefactorStepCompletion(command: string) {
+  const processResult = JSON.parse(await executeCommand(command));
+  return processResult.Status;
+}
+
+export async function stackRefactor(projRoot: string, category: RefactorCategory, bucketName: string) {
+  const { gen1ResourceIds, gen1ResourceDetails } = await getGen1ResourceDetails(projRoot, category);
+
+  const readmeFilePath = path.join(projRoot, '.amplify', 'migration', 'templates', category, 'MIGRATION_README.md');
+  const readmeContent = fs.readFileSync(readmeFilePath, 'utf-8');
+  const { step1Commands, step2commands, step3Commands } = getCommandsFromReadme(readmeContent);
+
+  await executeStep1(projRoot, step1Commands);
+  await executeStep2(projRoot, step2commands);
+  await executeStep3(projRoot, step3Commands, bucketName);
+
+  if (category === 'storage') await uncommentS3BucketLineFromBackendFile(projRoot);
+
+  await runGen2SandboxCommand(projRoot);
+
+  toggleEnvVariable('BUCKET_NAME', 'DELETE');
+  toggleEnvVariable('STACK_REFACTOR_ID', 'DELETE');
+
+  const { gen2ResourceIds, gen2ResourceDetails } = await getGen2ResourceDetails(projRoot, category);
+  assert.deepEqual(gen1ResourceIds, gen2ResourceIds);
+  assert.deepEqual(gen1ResourceDetails, gen2ResourceDetails);
```

Comment:
add assertion that these gen1ResourceIds no longer exists in gen1 stacks.

Reply:
The `gen2ResourceIds` are fetched from the `Outputs` after the final `npx ampx sandbox` which is the final step for stack refactor. I believe the resources can only be present in one stack and CFN doesn't allow it to be in two stacks. So, I think the assertion that gen1ResourceIds no longer exists in gen1 stacks gets handled there since we are making sure the ids are fetched from the outputs and not reusing `gen2ResourceIds `. Also, to do this assertion, I would have to make a separate call to get the nested stack name and other call for assertion.

Reply:
fair point!